### PR TITLE
feat(protocol-engine): Implement `CommandState.get_next_request()`

### PR DESCRIPTION
**Pull request by SyntaxColoring**
# Overview

Document, test, and implement `opentrons.protocol_engine.state.commands.get_next_request()`, which was formerly a `NotImplementedError`, in service towards #7808.

# Review requests

Mostly on testing.

# Risk assessment

None. Not production code, and this wasn't implemented before.

### CODE REVIEW

File: api/tests/opentrons/protocol_engine/state/test_command_state.py [hunk 1]
```diff
@@ -27,3 +54,110 @@ def test_state_store_handles_command(store: StateStore, now: datetime) -> None:
...
+        ("command-id-1", command_c), ("command-id-2", command_b)
+    ]
+
+
+# todo(mm, 2021-06-10): Not sure if this is the right way to initialize the queue
+# of commands for the purposes of this test.
+#
+# Options:
+#
+# * Manually initalize ._commands_by_id? ._commands_by_id is private (sort of), and
```
Comment (mcous) [inline api/tests/opentrons/protocol_engine/state/test_command_state.py]: Do this option. `_commands_by_id` shouldn't be private in this way, and is made public in #7882, so implementing the test this way will be the most similar

File: api/src/opentrons/protocol_engine/state/commands.py [hunk 2]
```diff
@@ -1,34 +1,52 @@
...
 class CommandState:
     """Command state and getters."""
 
-    _commands_by_id: Dict[str, CommandType]
+    _commands_by_id: typing.OrderedDict[str, CommandType]
 
     def __init__(self) -> None:
         """Initialize a CommandState instance."""
-        self._commands_by_id = {}
+        self._commands_by_id = collections.OrderedDict()
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/state/commands.py:17]: Is this change necessary? [According to the Python docs](https://docs.python.org/3/library/stdtypes.html#dict), plain `dict` is guarenteed to be in insertion order as of Python 3.7.

> Dictionaries preserve insertion order. Note that updating a key does not affect the order. Keys added after deletion are inserted at the end.
>
> ```
> >>> d = {"one": 1, "two": 2, "three": 3, "four": 4}
> >>> d
> {'one': 1, 'two': 2, 'three': 3, 'four': 4}
> >>> list(d)
> ['one', 'two', 'three', 'four']
> >>> list(d.values())
> [1, 2, 3, 4]
> >>> d["one"] = 42
> >>> d
> {'one': 42, 'two': 2, 'three': 3, 'four': 4}
>>> del d["two"]
>>> d["two"] = None
>>> d
> {'one': 42, 'three': 3, 'four': 4, 'two': None}
> ```
>
> Changed in version 3.7: Dictionary order is guaranteed to be insertion order. This behavior was an implementation detail of CPython from 3.6.

File: api/tests/opentrons/protocol_engine/state/test_command_state.py [hunk 3]
```diff
@@ -1,16 +1,43 @@
...
+            loadName=f"load-name-{i}",
+            namespace="opentrons-test",
+            version=1,
+            location=DeckSlotLocation(slot=DeckSlotName.SLOT_1),
+            labwareId=None
+        )
+    requests = [request(i) for i in range(n)]
+    for a, b in itertools.combinations(requests, 2):
+        # Testing the test. If these accidentally compare equal real assertions in this
+        # module (e.g. to compare iteration order) could trivially succeed.
```
Comment (mcous) [inline api/tests/opentrons/protocol_engine/state/test_command_state.py]: I get that this is technically true, but given our use of dataclasses (or Pydantic classes in the simplification work), this feels excessive. It "improves" the accuracy of the tests in a pedantic sense, but it harms the overall grokability (and therefore maintainability) of the test suite, in my opinion

File: api/tests/opentrons/protocol_engine/state/test_command_state.py [hunk 4]
```diff
@@ -27,3 +54,110 @@ def test_state_store_handles_command(store: StateStore, now: datetime) -> None:
     store.handle_command(cmd, command_id="unique-id")
 
     assert store.commands.get_command_by_id("unique-id") == cmd
+
+
+def test_command_state_preserves_handle_order(  # noqa:D103
```
Comment (mcous) [inline api/tests/opentrons/protocol_engine/state/test_command_state.py]: This is a good test to have 👍

File: api/tests/opentrons/protocol_engine/state/test_command_state.py [hunk 5]
```diff
@@ -27,3 +54,110 @@ def test_state_store_handles_command(store: StateStore, now: datetime) -> None:
...
+    store: StateStore, now: datetime
+) -> None:
+    unique_commands = [
+        PendingCommand[LoadLabwareRequest, LoadLabwareResult](
+            created_at=now,
+            request=r
+        )
+        for r in _make_unique_requests(3)
```
Comment (mcous) [inline api/tests/opentrons/protocol_engine/state/test_command_state.py]: I think this fixture function might be doing a little too much. Is this easier to read as an unrolled loop, without using actual commands?

```python
command_a = PendingCommand(
  created_at=datetime(year=2021, month=1, day=1),
  request=BaseModel()
)
command_b = PendingCommand(
  created_at=datetime(year=2022, month=2, day=2),
  request=BaseModel()
)
command_c = PendingCommand(
  created_at=datetime(year=2023, month=3, day=3),
  request=BaseModel()
)

store.handle_command(command_a, "command-id-1")  # type: ignore[arg-type]
store.handle_command(command_b, "command-id-2")  # type: ignore[arg-type]

assert store.commands.get_all_commands() == [
  ("command-id-1", command_a),
  ("command-id-2", command_b),
]

store.handle_command(command_c, "command-id-1")  # type: ignore[arg-type]

assert store.commands.get_all_commands() == [
  ("command-id-1", command_c),
  ("command-id-2", command_b),
]
```

File: api/tests/opentrons/protocol_engine/state/test_command_state.py [hunk 6]
```diff
@@ -27,3 +54,110 @@ def test_state_store_handles_command(store: StateStore, now: datetime) -> None:
...
+# * Add the commands through store.handle_command()? That would make this
+#   test partially redundant with test_state_store_handles_command(), which seems
+#   Bad.
+# * Mock out CommandState.get_all_commands() and avoid the need to initialize the
+#   queue of commands at all? That would mean CommandState is partially mocked out,
+#   which seems Bad.
+def test_get_next_request_returns_first_pending(  # noqa: D103
+    store: StateStore, now: datetime
+) -> None:
+    running_command = RunningCommand[LoadLabwareRequest, LoadLabwareResult](
```
Comment (mcous) [inline api/tests/opentrons/protocol_engine/state/test_command_state.py]: I think it was a mistake for me to have written these tests with `LoadLabware` commands. I think it's easier to read and write these tests if they just use `RunningCommand` etc. classes directly and shut mypy up where needed.

You can see [newer command state tests in 7882](https://github.com/Opentrons/opentrons/pull/7882/files#diff-16b3d16ccfbffca411f372719c918b34aba7cb834073780d4f3e0283a234b944) where I do this

File: api/tests/opentrons/protocol_engine/state/test_command_state.py [hunk 7]
```diff
@@ -27,3 +54,110 @@ def test_state_store_handles_command(store: StateStore, now: datetime) -> None:
...
+        request=_make_request()
+    )
+    pending_command = PendingCommand[LoadLabwareRequest, LoadLabwareResult](
+        created_at=now,
+        request=_make_request()
+    )
+
+    store.handle_command(running_command, "command-id-1")
+    store.handle_command(pending_command, "command-id-2")
+    # Skips running_command even though it came first.
```
Comment (mcous) [inline api/tests/opentrons/protocol_engine/state/test_command_state.py]: I think this test would be a little more through if it had one of every available non-failed status ahead of pending

File: api/tests/opentrons/protocol_engine/state/test_command_state.py [hunk 8]
```diff
@@ -27,3 +54,110 @@ def test_state_store_handles_command(store: StateStore, now: datetime) -> None:
...
+            created_at=now,
+            started_at=now,
+            failed_at=now,
+            error=ProtocolEngineError(),
+            request=_make_request()
+        ),
+        "failed-command-id"
+    )
+
+    # todo(mm, 2021-06-11): We should throw a completed command in here too, but I'm
```
Comment (mcous) [inline api/tests/opentrons/protocol_engine/state/test_command_state.py]: See other comment about not using real commands, and instead using values that are simply command classes of the correct shape and value types

**Review by mcous — COMMENTED** [review]
Feedback on tests, mostly

File: api/src/opentrons/protocol_engine/state/commands.py (hunk 2 above)
Comment (SyntaxColoring) [reply api/src/opentrons/protocol_engine/state/commands.py:17]: Nope, not technically. It's just for clarity.

File: api/tests/opentrons/protocol_engine/state/test_command_state.py (hunk 4 above)
Comment (SyntaxColoring) [inline api/tests/opentrons/protocol_engine/state/test_command_state.py]: I'm omitting docstrings and `#noqa:D103`ing these test functions because the *function names* are what show up in the PyTest summary, so it seems like I should invest all my descriptive energy into making *those* clear and readable.

If we think this is weird and inconsistent, I'm happy to add docstrings now and reserve this discussion for another day.

Comment (mcous) [reply api/tests/opentrons/protocol_engine/state/test_command_state.py]: I like the docstrings because I much prefer BDD-style `describe` and `it` API of [mocha](https://mochajs.org/#bdd) (also jasmine, [rspec](https://rspec.info/documentation/3.10/rspec-core/), etc.) as opposed to the flatter `test_xyz` style pytest uses. I've been using the docstrings as a way to write the `it("should behave this way", () => {` message that my brain craves

Comment (SyntaxColoring) [reply api/tests/opentrons/protocol_engine/state/test_command_state.py]: Yeah, I like that too.

When I try that, I feel like I frequently collide with the unfortunate combination of:

* The first docstring sentence must fit on a single line.
* A single line must be less than 80 (I think?) characters.

Like, I want the `"it should..."` sentence to be an *elaboration* of whatever the function name `test_command_state_preserves_handle_order` means...but there's not enough room for that, so it becomes a restatement instead.

Maybe I need to bite the verbosity bullet and do a cascade like:

```python
def test_command_state_preserves_handle_order():
    """It should preserve handle order.                                     <-- [This bit bugs me.]

    It should return commands in the order they were initially added.
    Replacing the command at an existing ID should keep the original
    command's place in the ordering.
    """
```

Comment (mcous) [reply api/tests/opentrons/protocol_engine/state/test_command_state.py]: With that I think the first sentence of your description is better, and this would be enough for me:

```py
def test_command_state_preserves_order() -> None:
   """It should return commands in the order they are added."""
    # ...
```

File: api/tests/opentrons/protocol_engine/state/test_command_state.py (hunk 5 above)
Comment (SyntaxColoring) [reply api/tests/opentrons/protocol_engine/state/test_command_state.py]: I think I like rolling the loop into a function because it lets us name which trait in this triplet of commands is important. For example, unrolling the loop means we have to write `PendingCommand` (I think?), which feels overly specific. We care that they're unique, and nothing else.

Also, at the very least, I should simplify the usage to something like:

```python
[command_a, command_b, command_c] = _make_unique_commands(3)
```

File: api/tests/opentrons/protocol_engine/state/test_command_state.py (hunk 3 above)
Comment (SyntaxColoring) [reply api/tests/opentrons/protocol_engine/state/test_command_state.py]: To clarify, this isn't meant to test that dataclasses implement `!=` properly. It's meant to test that I'm constructing each object with properly unique values. And that I'm not, for example, doing something like omitting the `f` before the f-string.

Does that change your opinion?

Comment (mcous) [reply api/tests/opentrons/protocol_engine/state/test_command_state.py]: After reading through it all again, I think a more accurate way to put my thoughts into words is: "the fact that this fixture needs a self-test means that this fixture is too complicated, and I think the tests would be easier to read and reason with without the fixture"

File: api/tests/opentrons/protocol_engine/state/test_command_state.py (hunk 5 above)
Comment (mcous) [reply api/tests/opentrons/protocol_engine/state/test_command_state.py]: What if you used the other value fixtures?

```py
def test_command_state_preserves_handle_order(
    pending_command: PendingCommand,
    running_command: RunningCommand,
    completed_command: CompletedCommand,
) -> None:
    store.handle_command(running_command, "command-id-1")
    store.handle_command(pending_command, "command-id-2")

    assert store.commands.get_all_commands() == [
        ("command-id-1", running_command),
        ("command-id-2", pending_command),
    ]

    store.handle_command(completed_command, "command-id-1")

    assert store.commands.get_all_commands() == [
        ("command-id-1", completed_command),
        ("command-id-2", pending_command),
    ]
```

File: api/tests/opentrons/protocol_engine/state/test_command_state.py (hunk 3 above)
Comment (SyntaxColoring) [reply api/tests/opentrons/protocol_engine/state/test_command_state.py]: Moot with removal of `make_unique_requests()`. See https://github.com/Opentrons/opentrons/pull/7936#discussion_r650312525.

File: api/tests/opentrons/protocol_engine/state/test_command_state.py (hunk 5 above)
Comment (SyntaxColoring) [reply api/tests/opentrons/protocol_engine/state/test_command_state.py]: Definitely better, but ultimately I think I have the same objection. `pending_command`, `running_command`, and `completed_command` seem lower than the most meaningful level of abstraction for this particular test.

For now, I'm moving forward with your suggestion, plus a comment pointing out that the commands are just an arbitrary triplet that compares `!=` to each other.

File: api/tests/opentrons/protocol_engine/state/test_command_state.py (hunk 4 above)
Comment (SyntaxColoring) [reply api/tests/opentrons/protocol_engine/state/test_command_state.py]: Moving forward with docstrings similar to that latest suggestion. Though it still feels a little unsatisfying to me, since it's not "semantically different" enough from the function name, if that makes sense.

**Review by mcous — APPROVED** [review]

**Review by sanni-t — COMMENTED** [review]
👀
